api/v1alpha1/bmcjob_types.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 Tinkerbell.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BMCJobConditionType represents the condition of the BMC Job.
+type BMCJobConditionType string
+
+const (
+	// JobCompleted represents successful completion of the BMC Job tasks.
+	JobCompleted BMCJobConditionType = "Completed"
+	// JobFailed represents failure in BMC job execution.
+	JobFailed BMCJobConditionType = "Failed"
+	// JobTaskCompleted represents successful completion of a BMC Job task.
+	JobTaskCompleted BMCJobConditionType = "TaskCompleted"
```

Comment:
True. 
So my intention was to use this in the conditions for the `BMCJobStatus`. 
Would this be a different `Conditions` slice all together then? I was thinking to have task status as part of job status itself.

Reply:
We don't actually have a type for a task, its just a struct. If there are multiple tasks how does this work? It seems to be synonymous with `JobCompleted`.

Reply:
What status does a task have and why would that be modeled on a job as a state? You could have `Tasks` and `CompletedTasks` moving tasks from one slice to the next so we know how far the job got but I'm not sure you'd want much more than that on the job type.

Reply:
My thought was, we would append to the `.Status.Conditions` slice. Like
```yaml
[ 
  {
    type: "TaskCompleted",
    message: "Power On task completed successfully"
  },
  {
    type: "TaskCompleted",
    message: "Boot Device PXE task completed successfully"
  },
  {
    type: "JobCompleted"
  }
]
```

This way we can track the individual tasks, so if a user wants to know what tasks have ran and even which failed.
That is why I kept it as a `BMCJobConditionType`, since tasks are part of a job.

Reply:
I see a few issues with that. (1) we're creating a dynamic number of conditions that could exist on a job; (2) consumers need to encode/cater for the dynamic state count in their logic; (3) we're encoding information in a free form string about specific, finite tasks because all tasks are represented with a single type - thats different to opaque error messages.

I can't decide if that's really bad but my gut tells me its not great - will need to bake on it. What I can say with confidence is that it increase the complexity of the API and I'm not sure its justifying the added complexity.

Reply:
Right, that's a good point. 
My motivation was to have a way to track individual tasks without too much complexity. 
We could model it with a separate slice, so it does not interfere with the job conditions itself. Example

```go
type BMCJobStatus struct {
  TaskConditions []TaskCondition `json:"taskConditions,omitempty"`
}

type TaskCondition struct {
  Type TaskConditionType `json:"type"`
  Message string `json:"message,omitempty"`
}

type TaskConditionType string

const (
  TaskFailed TaskConditionType = "TaskFailed"
  PowerTaskCompleted TaskConditionType = "PowerTaskCompleted"
  ...
)
```

This way consumers can look at the job `conditions` with limited complexity and look at the `TaskConditions` (we could also call it `TaskStatuses` or `TaskResults`) to see what passed/failed.